ext/standard/string.c

```diff
@@ -1035,7 +1035,7 @@ PHP_FUNCTION(implode)
 
 	if (pieces == NULL) {
 		if (arg1_array == NULL) {
-			zend_type_error("%s(): Argument #1 ($array) must be of type array, string given", get_active_function_name());
+			zend_type_error("%s(): Argument #1 ($array) must be of type array, null given", get_active_function_name());
```

Comment:
This error message is for argument 1. Your test tests argument 2. Implode has two distinct signatures:

```php
 implode(string $separator, array $array): string
implode(array $array): string
```

When argument 2 is not set (or null) argument 1 must be an array.

Reply:
You are correct. My mistake. The PR is indeed incorrect.

But if argument 2 is `null`, I wouldn't expect the second signature to be used. I would expect to see the type error.

Reply:
:shrug: PHP doesn't really support overloading. It was decided to use the type of arg2 as a deciding factor to switch between the signatures rather than the number of args. Changing this would be a subtle BC break.

Reply:
I understand, but don't you think this is worth adding (even if it's for future versions)?

I encountered this problem in the following case:
```
echo implode(', ', functionWithNullableArrayReturn($parameters));
```

The error was really confusing to me and looked like a bug.

Reply:
If we want to change this, it's better to split the paths as in https://github.com/php/php-src/pull/12583.

Reply:
Yep, it does look better. I'll work on it. :-)

Reply:
I started splitting it but now I get the following error:

> Argument #1 ($separator) must be of type string, string given

I'm having trouble understanding what may be causing it. Did I get something wrong? 😳

Reply:
There's an additional problem that the arguments for 1 and 2 are essentially flipped. ZPP looks up parameter names from the arg info. So this approach might not be  better after all.

Reply:
Sorry, I didn't understand. What do you mean? 😳

Reply:
Ah, it should be `ZEND_PARSE_PARAMETERS_START(2, 2)`. That fixed the issue.

I still don't understand what you meant, but now the code works. haha

If you're ok with this implementation, I'll fix the failing tests. :-D